controller: client_steering_task: wait for disconnect and connect

arnout opened the pull request:
The steering task is considered successful when the connected event for
the STA is received. However, we were not waiting for that event, so if
the disconnected event arrives before the connected event, we would
first jump to the FINALIZE state and report that steering has failed.

Add STA_CONNECTED to the events that we're waiting for.

Add test to make sure this works.

Fixes #664

rmelotte commented:
What happens if the client first connects to the next AP before disconnecting from the current one?
Will the connect event still be handled or will the steering task wait until another connect event occurs?

I think both cases should be tested by `test_flows.sh`: 
- the client first connects to the next AP, then disconnects from the current one.
- the client first disconnects from the current one, then connects to the next one. 

We need to know (as in: test in CI) what happens in both cases: will the steering task hang/fail/succeed?

arnout committed:
test_flows: client_steering_dummy: check for correct completion

Currently only "steering successful" is checked. It turns out, however,
that sometimes the connect or disconnect message doesn't get properly
handled by the steering task. To detect this case:
- add a small sleep between the connect and disconnect, to make sure the
test is predictable;
- add a check that the disconnect was correctly handled by also checking
for the "disconnected after successful steering" message.

Note that this change causes the test to fail reliably.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
controller: client_steering_task: wait for disconnect and connect

The steering task is considered successful when the connected event for
the STA is received. However, we were not waiting for that event, so if
the disconnected event arrives before the connected event, we would
first jump to the FINALIZE state and report that steering has failed.

Add STA_CONNECTED to the events that we're waiting for.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout commented:
> I think both cases should be tested by `test_flows.sh`:
> 
>     * the client first connects to the next AP, then disconnects from the current one.
> 
>     * the client first disconnects from the current one, then connects to the next one.

You're completely right. I updated the test to reliably test one of the two cases, and now if tails always even after the fix :-). So that will have to be fixed as well. Marking as don't merge until that is done.

arnout committed:
controller: always inform steering task of disconnect

The disconnect event was only passed to the steering task if the
handoff_flag was false, i.e. if there was no steering ongoing. This will
obviously never work. Move the steering task event out of the handoff
condition.

While we're at it, only push it to the steering task if the steering
task is running - otherwise push_event will print an error message.

Also while we're at it, update the comments to clarify what is going on.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Client steering doesn't correctly report success](https://trello.com/c/0R3DY5N3/66-client-steering-doesnt-correctly-report-success)

rmelotte reviewed (approved):
I created a new task to test "connect before disconnect" (#685) so that we can already merge this one.

arnout commented:
I rebased and re-tested each individual commit. The test now consistently fails after the first commit, still fails after the second commit, and succeeds after the third one.
